4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py

```diff
@@ -0,0 +1,283 @@
+            study = optuna.create_study(direction="minimize")
+            study.optimize(lambda trial: objective_tc(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)
+            # Train the best BetaTCVAE model and extract z matrix
+            best_trial = study.best_trial
+            model = BetaTCVAE(input_dim=train_data.shape[1], latent_dim=num_components, beta=best_trial.params['beta'])
+            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)
+            optimizer = get_optimizer_tc(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])
+            train_tc_vae(model, train_loader, optimizer, best_trial.params['epochs'])
+            
+            # Extract weight matrix
+            weight_matrix_df = tc_weights(model, weight_data)
+            weight_matrix_df.rename(columns={0: 'genes'}, inplace=True)
+        
+        elif model_name == "vanillavae":
+            # Optuna optimization for VanillaVAE
+            study = optuna.create_study(direction="minimize")
+            study.optimize(lambda trial: objective_vvae(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)
+            
+            # Train the best BetaTCVAE model and extract z matrix
+            best_trial = study.best_trial
+            model = VanillaVAE(input_dim=train_data.shape[1], latent_dim=num_components)
+            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)
+            optimizer = get_optimizer_vvae(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])
+            train_vvae(model, train_loader, optimizer, best_trial.params['epochs'])
+            
+            # Extract weight matrix
+            weight_matrix_df = vanilla_weights(model, weight_data)
+            weight_matrix_df.rename(columns={0: 'genes'}, inplace=True)
+        # If weight_matrix is generated, proceed to GSEA and append to combined dataframe
+        if weight_matrix_df is not None:
+            print(f"Running GSEA for {model_name}")
+            gsea_results_df = perform_gsea(weight_matrix_df, model_name, num_components)
+            combined_results_df = pd.concat([combined_results_df, gsea_results_df], ignore_index=True)
+            
+
+
+# In[8]:
+
+
+# Save the combined dataframe to a file
+final_output_file = output_dir / "combined_z_matrix_gsea_results.parquet"
+combined_results_df.to_parquet(final_output_file, index=False)
+
+print(f"Saved final combined z_matrix and GSEA results to {final_output_file}")
+
+#Save as CSV for R 
+csv_output_file = output_dir / "combined_z_matrix_gsea_results.csv"
+combined_results_df.to_csv(csv_output_file, index=False)
+
+
+# In[9]:
+
+
+combined_results_df.sort_values(by='gsea es score', key=abs, ascending = False).head(50)
```

Comment:
I must have missed this earlier in the notebook, but try to avoid using spaces in column names (make it `gsea_es_score`). Spaces remove functionality in pandas (specifically, column referencing with dot notation [e.g., `combined_results_df.gsea es score` does not work])

Reply:
Fixed!